flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/metrics/MetricStore.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.runtime.webmonitor.metrics;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Nested data-structure to store metrics.
+ *
+ * This structure is not thread-safe.
+ */
+public class MetricStore {
+	private static final Logger LOG = LoggerFactory.getLogger(MetricStore.class);
+
+	JobManagerMetricStore jobManager = new JobManagerMetricStore();
+	Map<String, TaskManagerMetricStore> taskManagers = new HashMap<>();
+	Map<String, JobMetricStore> jobs = new HashMap<>();
+
+	/**
+	 * Adds a metric to this MetricStore.
+	 *
+	 * @param name  the metric identifier
+	 * @param value the metric value
+	 */
+	public void add(String name, Object value) {
```

Comment:
I think the object array representation (with the contained objects) is not as concise as the byte buffer representation (using your own serialization) because the standard Java serialization will write for every object the class name to identify it.

But again this falls imo into the category of premature optimization.